### ISSUE TITLE
Support linking unnamed temp files into filesystem on Linux

### ISSUE THREAD

**Issue by e60e0e11-52cd-4cd6-9601-15baf1eb36d6** [python/cpython#89341]
BPO | [45178](https://bugs.python.org/issue45178)
--- | :---

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-09-12.23:45:46.424>
labels = ['type-feature', 'library', '3.11']
title = 'Support linking unnamed temp files into filesystem on Linux'
updated_at = <Date 2021-09-18.01:22:30.534>
user = 'https://bugs.python.org/WGH'
```

bugs.python.org fields:
```python
activity = <Date 2021-09-18.01:22:30.534>
actor = 'terry.reedy'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2021-09-12.23:45:46.424>
creator = 'WGH'
dependencies = []
files = []
hgrepos = []
issue_num = 45178
keywords = []
message_count = 1.0
messages = ['401676']
nosy_count = 1.0
nosy_names = ['WGH']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'test needed'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue45178'
versions = ['Python 3.11']
```

</p></details>

### Linked PRs
* gh-136302

**Comment by jbosboom** [comment]
#81793 was fixed, so this is now possible when `/proc` is accessible:

```
import os
fd = os.open('/tmp', os.O_RDWR | os.O_TMPFILE, 0o600)
os.write(fd, b'abcdef\n')
os.link(f'/proc/self/fd/{fd}', '/tmp/foo', follow_symlinks=True)
```

(There's no way to use `AT_EMPTY_PATH`, but that's only relevant when `/proc` is not accessible but the program has `CAP_DAC_READ_SEARCH`, which seems like an uncommon combination.)

`tempfile.TemporaryFile` still passes `O_EXCL` unconditionally.

**Comment by serhiy-storchaka** [comment]
I tried to implement this feature, but since it requires the CAP_DAC_READ_SEARCH capability, it usually only works for the root. So I doubt the value of this feature.

cc @vstinner

**Comment by serhiy-storchaka** [comment]
It is difficult to check if the process has the CAP_DAC_READ_SEARCH capability.

* Use `getpcaps` and then decode its cryptic output.
* Search the "CapEff" field in `/proc/{pid}/status`, then decode the numerical value with `capsh --decode`.
* Implement .

It depends on `/proc` or external commands or requires writing complex C code. But if you have working `/proc`, you do not need this feature. `os.link(f'/proc/self/fd/{fd}', linkname)` is superior way, it works even without the CAP_DAC_READ_SEARCH capability.

**Comment by jbosboom** [comment]
Sorry if I wasn't clear.  I mentioned `AT_EMPTY_PATH` only for completeness, not as a feature request.  I agree it's rarely useful.

**Comment by serhiy-storchaka** [comment]
I agree with you that that feature could only be useful in very uncommon situation. In most cases creating a link to `f'/proc/self/fd/{fd}'` is more preferable.

Recent [change](https://github.com/torvalds/linux/commit/42bd2af5950456d46fdaa91c3a8fb02e680f19f5) in the Linux kernel made it more useful. But the kernel on my computer is older, so it is not easy to me to test it for unprivileged user. If anyone is interested, you can continue my work (you need to add documentation and update the test that should check the Linux version).
